File: admin/backend/api/v1/editor/routes/git.py
from __future__ import annotations

from flask import jsonify, request

from admin.backend.api.responses import error_response
from admin.backend.api.v1.editor import editor_bp, json_body, with_workspace


@editor_bp.get("/git/status")
@with_workspace
def git_status(ws):
    return jsonify(ws.git.status())


@editor_bp.get("/git/blame")
@with_workspace
def git_blame(ws):
    return jsonify({"lines": ws.git.blame(ws.relative(request.args.get("path", "")))})


@editor_bp.get("/git/diff")
@with_workspace
def git_diff(ws):
    return jsonify(ws.git.diff_gutters(ws.relative(request.args.get("path", ""))))


@editor_bp.get("/git/show")
@with_workspace
def git_show(ws):
    return jsonify({"content": ws.git.show_head(ws.relative(request.args.get("path", "")))})


@editor_bp.get("/git/branches")
@with_workspace
def git_branches(ws):
    return jsonify(ws.git.branches())


@editor_bp.post("/git/checkout")
@with_workspace
def git_checkout(ws):
    body = json_body()
    if not body.get("branch"):
        return error_response("bad_request", "Branch is required.", 400)
    return _git_result(*ws.git.checkout(body["branch"], bool(body.get("create"))))


@editor_bp.post("/git/commit")
@with_workspace
def git_commit(ws):
    body = json_body()
    if not (body.get("message") or "").strip():
        return error_response("bad_request", "Commit message is required.", 400)
    return _git_result(*ws.git.commit(body["message"], bool(body.get("all"))))


@editor_bp.get("/git/log")
@with_workspace
def git_log(ws):
    skip = request.args.get("skip", type=int) or 0
    limit = request.args.get("limit", type=int) or 50
    if limit <= 0 or limit > 200:
        limit = 50
    return jsonify(ws.git.log(skip, limit))


@editor_bp.get("/git/commit-info")
@with_workspace
def git_commit_info(ws):
    sha = request.args.get("sha", "")
    if not ws.git.is_sha(sha):
        return error_response("bad_request", "Invalid commit sha.", 400)
    info = ws.git.commit_info(sha)
    if info is None:
        return error_response("not_found", "Commit not found.", 404)
    return jsonify(info)


@editor_bp.get("/git/commit-diff")
@with_workspace
def git_commit_diff(ws):
    sha = request.args.get("sha", "")
    if not ws.git.is_sha(sha):
        return error_response("bad_request", "Invalid commit sha.", 400)
    rel = ws.relative(request.args.get("path", ""))
    return jsonify(ws.git.commit_file(sha, rel))


@editor_bp.post("/git/stage")
@with_workspace
def git_stage(ws):
    return _git_path_op(ws, ws.git.stage)


@editor_bp.post("/git/unstage")
@with_workspace
def git_unstage(ws):
    return _git_path_op(ws, ws.git.unstage)


@editor_bp.post("/git/discard")
@with_workspace
def git_discard(ws):
    path = (json_body().get("path") or "").strip()
    if not path:
        return error_response("bad_request", "Path is required.", 400)
    return _git_result(*ws.git.discard(ws, ws.relative(path)))


@editor_bp.post("/git/push")
@with_workspace
def git_push(ws):
    return _git_result(*ws.git.push(bool(json_body().get("force"))))


@editor_bp.post("/git/pull")
@with_workspace
def git_pull(ws):
    return _git_result(*ws.git.pull())


def _git_path_op(ws, op):
    path = (json_body().get("path") or "").strip()
    if not path:
        return error_response("bad_request", "Path is required.", 400)
    return _git_result(*op(ws.relative(path)))


def _git_result(ok: bool, message: str):
    response = jsonify({"ok": ok, "message": message})
    response.status_code = 200 if ok else 409
    return response

File: admin/frontend/editor/src/api/git.js
import { request, body, bodyAny } from './client'

export const gitApi = {
  status: () => request.get('git/status').then((r) => body(r, { repo: false })),
  diff: (path) =>
    request.get('git/diff', { searchParams: { path } }).then((r) => body(r, { added: [], modified: [], deleted: [] })),
  show: (path) => request.get('git/show', { searchParams: { path } }).then((r) => body(r, { content: '' })),
  blame: (path) => request.get('git/blame', { searchParams: { path } }).then((r) => body(r, { lines: [] })),
  branches: () => request.get('git/branches').then((r) => body(r, { current: '', branches: [] })),
  log: (skip = 0, limit = 50) =>
    request.get('git/log', { searchParams: { skip, limit } }).then((r) => body(r, { repo: false, commits: [], more: false })),
  commitInfo: (sha) => request.get('git/commit-info', { searchParams: { sha } }).then((r) => body(r)),
  commitDiff: (sha, path) =>
    request.get('git/commit-diff', { searchParams: { sha, path } }).then((r) => body(r, { old: '', new: '' })),

  checkout: (branch, create) => request.post('git/checkout', { json: { branch, create: !!create } }).then(bodyAny),
  commit: (message, all) => request.post('git/commit', { json: { message, all: !!all } }).then(bodyAny),
  stage: (path) => request.post('git/stage', { json: { path } }).then(bodyAny),
  unstage: (path) => request.post('git/unstage', { json: { path } }).then(bodyAny),
  discard: (path) => request.post('git/discard', { json: { path } }).then(bodyAny),
  push: (force) => request.post('git/push', { json: { force: !!force } }).then(bodyAny),
  pull: () => request.post('git/pull', { json: {} }).then(bodyAny),
}
